Docs: Plugin doc review for chunk 1-A (#67070)
grafana/grafana: docs/sources/developers/plugins/build-a-logs-data-source-plugin.md

# Build a logs data source plugin

Data logs are defined as data stored over a period of time in different systems or environments for the purpose of tracking certain events.  

> Grafana data source plugins support metrics , logs, and other data types. The steps to build a logs data source plugin are largely the same as for a metrics data source, but there are a few differences. 

Reviewer: there is an extra space there: `metrics, logs, and other data types`

Author: fixed in latest commit

> # Build a logs data source plugin

Reviewer 2: should this be a sub page of Create a Grafana plugin?

Author: Yes, it will move there in phase 2.

> Data logs are defined as data stored over a period of time in different systems or environments for the purpose of tracking certain events.  

Reviewer 2: I've never come across the term "Data logs" and I'm not sure this definition is scoped enough to logs... I could say the same for metrics and alerts. Again I'm not sure we need to define logs

Author: Issue is fixed in latest commit.

> Grafana data source plugins support metrics, logs, and other data types. The steps to build a logs data source plugin are largely the same as for a metrics data source, but there are a few differences. 

Reviewer 2: Instead of defining logs above, we could expand on _why_ there are the differences required compared to a metrics data source? (I cannot provide a suitable explanation here, but I'm sure someone else can). Alternatively we could end with "...but there are a few difference which we will explain in this guide."

Author: I've gone with your latter suggestion for now. When we have more background info, we can add it then. Created https://github.com/grafana/grafana/issues/67428

# Build a logs data source plugin

> This guide explains how to build a logs data source plugin. 

Reviewer 3: Consider deleting this sentence duplicates the page title.

Author: Done. I will check in this fix soon.

Author: Check in fix, available for review.

> Data logs are defined as data stored over a period of time in different systems or environments for the purpose of tracking a variety of events.  

Reviewer 3: Does it work to say just "events" instead of "a variety of events"?  Does using the word "variety" imply there will be a list of events following? 
```suggestion
Data logs are defined as data stored over a period of time in different systems or environments for the purpose of tracking events.  
```

Author: I will check in fix soon, changing to "certain events".

Author: Check in fix, available for review.

> Grafana data source plugins support both metrics and log data. The steps to build a logs data source plugin are largely the same as for a metrics data source, but there are a few differences. 

Reviewer: nit: Grafana now supports more types of data apart than metrics and logs so it's not just those two

Author: Fixed, will check in soon.

Author: Checked in fix, available for review.

## Before you begin

> This guide assumes that you're already familiar with how to [Build a data source plugin](/tutorials/build-a-data-source-plugin/) for metrics. We recommend that you review this material before continuing.

Reviewer 3: I like your callout to the other procedure as a reference and review before continuing.

Author: Checked in fix, available for review.

Reviewer 3: Looks good

## Add logs support to your data source

To add logs support to an existing data source, you need to:

1. Enable logs support
1. Construct the log data

> When these steps are done, then you can improve the user experience with one or more optional features.

Reviewer 3: Should you link to the optional features?

Author: Done, will check in soon

Congratulations, you just wrote your first logs data source plugin! Next, let's look at a couple of features that can further improve the experience for the user.

## Enhance your logs data source plugin with optional features

Add visualization type hints, labels, and other optional features to logs.

### Add a preferred visualization type hint to the data frame

> To make sure Grafana recognizes data as logs and shows logs visualization automatically in Explore you have to set `meta.preferredVisualisationType` to `'logs'` in the returned data frame. See [Selecting preferred visualisation section]({{< relref "add-support-for-explore-queries/#selecting-preferred-visualisation" >}})

Reviewer 3: ```suggestion
To make sure Grafana recognizes data as logs and shows logs visualization automatically in Explore, set `meta.preferredVisualisationType` to `'logs'` in the returned data frame. See [Selecting preferred visualisation section]({{< relref "add-support-for-explore-queries/#selecting-preferred-visualisation" >}})
```

Congratulations, you just wrote your first logs data source plugin! Next, let's look at a couple of features that can further improve the experience for the user.

> ## Enhance your logs data source plugin with optional features

Reviewer 3: The writer's toolkit suggests using ["Optional: ..."](https://github.com/grafana/writers-toolkit/blob/main/docs/static/templates/tutorial-template.md) in the heading. Would that be a useful change here? Also, there is only one subheading under the enhancements, so is there only one possible enhancements? If so, do we even need this heading? 

Also, consider adding a sentence or two that summarizes what these optional features are.

Author: (Optional) is not in my draft text

> ### Add a preferred visualisation type hint to the data frame

Reviewer 3: If this is the only optional feature, then maybe add Optional: to this heading and remove the previous Enhance your logs data source plugins... heading entirely.

> To make sure Grafana recognizes data as logs and shows logs visualization automatically in Explore you have do set `meta.preferredVisualisationType` to `'logs'` in the returned data frame. See [Selecting preferred visualisation section]({{< relref "add-support-for-explore-queries/#selecting-preferred-visualisation" >}})

Reviewer 3: ```suggestion
To make sure Grafana recognizes data as logs and shows logs visualization automatically in Explore, you have to set `meta.preferredVisualisationType` to `'logs'` in the returned data frame. Refer to the [Selecting preferred visualization section]({{< relref "add-support-for-explore-queries/#selecting-preferred-visualisation" >}}).
```

Author: Fixed, will check in soon

Author: Checked in fix, available for review.

> ## Add labels to your logs

Reviewer 2: Are these not optional enhancements?

Author: Yes. I removed "(Optional)" because the title of this section changed to "## Enhance your logs data source plugin with optional features" and I avoided redundancy. If this is confusing, let's change it though.

Reviewer 2: > because the title of this section changed to "## Enhance your logs data source plugin with optional features" and I avoided redundancy

That I get, but the description for the new **section** has "Add visualization type hints, labels, and other optional features to logs." but then as shown above, the labels content is not _within_ this section, it's its own section.

Author: Issue is fixed in latest commit.

> To help filter log lines, many log systems let you query logs based on metadata, or _labels_.

Reviewer 3: ```suggestion
Many log systems let you query logs based on metadata, or _labels_, to help filter log lines.
```

Author: good suggestion, will check in fix soon

Author: Checked in fix, available for review.

> Add labels to a stream of logs by setting the labels property on the Field.

Reviewer 3: Should the labels property be formatted `labels`?

## Extract detected fields from your logs

Add additional information about each log line by supplying more data frame fields.

If a data frame has more than one text field, then Grafana assumes the first field in the data frame to be the actual log line. Grafana treats subsequent text fields as [detected fields]({{< relref "../../explore/#labels-and-detected-fields" >}}).

> Although you can add any number of custom fields to your data frame, Grafana comes with a couple of dedicated fields: `levels` and `id`. Let's have a closer look at each one.

Reviewer 2: Suggestion: "Any number of custom fields can be added to your data frame; Grafana comes with two dedicated fields: levels and id."

Reviewer 2: Also suggest we extend the "Unique log lines" subheader to indicate that it is id, otherwise you lose the connection between saying which two you're going to explain.

Author: Issue is fixed in latest commit.

frame.add({ time: 1589189406480, content: 'unknown error', level: 'error' });
```

> #### 'id' for assigning unique identifiers to log lines

Reviewer 3: Is there a more descriptive heading we could use here? For example, if the goal is to improve support for dedupe by adding an 'id' field, should that information be incorporated into the heading? Or even "Use `id` to support log dedupe". What do you think is the main goal that this section is wanting to help the user accomplish?
